source/fe/mapping_fe_field.cc

```diff
@@ -1840,12 +1842,13 @@ transform_real_to_unit_cell (const typename Triangulation<dim,spacedim>::cell_it
   UpdateFlags update_flags = update_quadrature_points | update_jacobians;
   if (spacedim>dim)
     update_flags |= update_jacobian_grads;
-  std::unique_ptr<InternalData>
+  std::unique_ptr<typename Mapping<dim,spacedim>::InternalDataBase>
   mdata (get_data(update_flags,point_quadrature));
 
-  update_internal_dofs(cell, *mdata);
+  update_internal_dofs(cell, dynamic_cast<InternalData &>(*mdata));
```

Comment:
It wasn't there before, but we usually check that the pointer version `dynamic_cast` succeeds and use `Assert(dynamic_cast(...), ExcInternalError())` rather than using the reference version (which can throw).

Reply:
I had thought about this. The object is created in the line just above this, from a function in the same class. So we know (or can at least look up) the actual type of the object, and if it is not what we expect, we will abort the program through the exception that `dynamic_cast` throws. So I wasn't overly concerned about hitting a difficult-to-find and invisible bug: if something goes wrong, we'll know it.

But I don't feel strongly about it. An `Assert` call has the advantage of pointing out the error with the correct location, as opposed to having to trace back where the exception came from. I'll be happy to add the assertions -- will get to it later today.